cores/esp32/esp32-hal-uart.c

```diff
@@ -294,6 +314,10 @@ uint8_t uartPeek(uart_t* uart)
         return 0;
     }
     uint8_t c;
+    if ((uxQueueMessagesWaiting(uart->queue) == 0) && (uart->dev->status.rxfifo_cnt > 0))
```

Comment:
If queue is not zero, then you ignore the data in the fifo?

I would think you would always add the `rxfifo_cnt` to the `uartPeek()` result.
```c++
if (uart->dev->status.rxfifo_cnt >0)
```

Reply:
If queue is not zero, the peek result is the first in the queue -> no need to transfer from fifo to queue now. 
Alternative is in both read and peek and possible even in available to always transfer all fifo to queue. 
It's a design decision, not functional, I'll change if decided to do the other way.
Just let me know

Reply:
The code sequence where this could matter is:
```c++
int packet_length = 10;
while (Serial.peek() < packet_length){
  delay(5);
}
processPacket();
```
So, I'm waiting for a complete packet before I process it. But this loop would not pass until rxTimeout or rxFifioFull.

You could decide not to empty the fifo, but just add the rxFifoCount to the returned.

What do you think? 

Also, there is something weird with the rxFifoCount, do you see in the original ISR coded where the "empty" test is rxFifoCount OR  `(uart->dev->mem_rx_status.wr_addr != uart->dev->mem_rx_status.rd_addr)`?  
It looks like rxFifoCount can be zero and the FiFo is FULL? @me-no-dev ?
perhaps rxFifoCount is not always correct?

Chuck.

Reply:
Serial.peek() doesn't return the size of the queue but the fist char in the queue, Serial.available() does return the number of chars available.
So your example should be
```
int packet_length = 10;
while (Serial.available() < packet_length){
  delay(5);
}
processPacket();
```
And that is providing the correct answer. 

On the rxFifocount. 
I agree with you I expect(ed) that the loop 
while(uart->dev->status.rxfifo_cnt) {... } 
would do the job..
But because of the current code I kept it identical.